Main.hs

```diff
@@ -4,36 +4,85 @@ import Control.Monad (unless)
 import Text.Printf (printf)
 
 data Expr
+  = Num Double
+  | Sqrt Expr
+  | Add Expr Expr
+  | Sub Expr Expr
+  | Mul Expr Expr
+  | Div Expr Expr
+  | Pow Expr Expr
 
 instance Show Expr where
-  show = undefined
+  show (Num n) = show n
+  show (Sqrt e) = "sqrt(" ++ show e ++ ")"
+  show (Add a b) = "(" ++ show a ++ " + " ++ show b ++ ")"
+  show (Sub a b) = "(" ++ show a ++ " - " ++ show b ++ ")"
+  show (Mul a b) = "(" ++ show a ++ " * " ++ show b ++ ")"
+  show (Div a b) = "(" ++ show a ++ " / " ++ show b ++ ")"
+  show (Pow a b) = "(" ++ show a ++ " ^ " ++ show b ++ ")"
 
 instance Eq Expr where
-  (==) = undefined
+  (Num a) == (Num b) = a == b
+  (Sqrt a) == (Sqrt b) = a == b
+  (Add a1 b1) == (Add a2 b2) = a1 == a2 && b1 == b2
+  (Sub a1 b1) == (Sub a2 b2) = a1 == a2 && b1 == b2
+  (Mul a1 b1) == (Mul a2 b2) = a1 == a2 && b1 == b2
+  (Div a1 b1) == (Div a2 b2) = a1 == a2 && b1 == b2
+  (Pow a1 b1) == (Pow a2 b2) = a1 == a2 && b1 == b2
+  _ == _ = False
 
 data Error
+  = DivByZero Expr Expr
+  | NegativeSqrt Expr
 
 instance Show Error where
-  show = undefined
+  show (DivByZero a b) = "Division by zero: " ++ show a ++ " / " ++ show b
+  show (NegativeSqrt e) = "Square root of negative number: " ++ show e
 
 instance Eq Error where
-  (==) = undefined
+  (DivByZero _ _) == (DivByZero _ _) = True
+  (NegativeSqrt _) == (NegativeSqrt _) = True
+  _ == _ = False
```

Comment:
I'd argue that `DivByZero (Num 1) (Num 0)` and `DivByZero (Num 1) (Sub (Num 1) (Num 1))` are two different errors.